packages/tailwindcss/src/variants.ts

```diff
@@ -42,33 +54,66 @@ export class Variants {
   static(
     name: string,
     applyFn: VariantFn<'static'>,
-    { compounds, order }: { compounds?: boolean; order?: number } = {},
+    { compounds, order }: { compounds?: Compounds; order?: number } = {},
   ) {
-    this.set(name, { kind: 'static', applyFn, compounds: compounds ?? true, order })
+    this.set(name, {
+      kind: 'static',
+      applyFn,
+      compoundsWith: Compounds.Never,
+      compounds: compounds ?? Compounds.StyleRules,
+      order,
+    })
   }
 
   fromAst(name: string, ast: AstNode[]) {
-    this.static(name, (r) => {
-      let body = structuredClone(ast)
-      substituteAtSlot(body, r.nodes)
-      r.nodes = body
+    let selectors: string[] = []
+
+    walk(ast, (node) => {
+      if (node.kind !== 'rule') return
+      if (node.selector === '@slot') return
+      selectors.push(node.selector)
     })
+
+    this.static(
+      name,
+      (r) => {
+        let body = structuredClone(ast)
+        substituteAtSlot(body, r.nodes)
+        r.nodes = body
+      },
+      {
+        compounds: compoundsForSelectors(selectors),
+      },
+    )
   }
 
   functional(
     name: string,
     applyFn: VariantFn<'functional'>,
-    { compounds, order }: { compounds?: boolean; order?: number } = {},
+    { compounds, order }: { compounds?: Compounds; order?: number } = {},
   ) {
-    this.set(name, { kind: 'functional', applyFn, compounds: compounds ?? true, order })
+    this.set(name, {
+      kind: 'functional',
+      applyFn,
+      compoundsWith: Compounds.Never,
+      compounds: compounds ?? Compounds.StyleRules,
+      order,
+    })
   }
 
-  compound(
+  compoundWith(
```

Comment:
If all compound variants need to specify their "with" do we actually need to change this name? I liked how before these methods were named as sort of factory functions for the variant kind, like `static`, `functional`, and `compound`. I don't think we want to think of these as "compoundWith" variants do we, vs. still just "compound"?